lib/ex_unit/test/ex_unit/case_test.exs

```diff
@@ -155,18 +155,70 @@ defmodule ExUnit.CaseTest.TmpDir do
 
   @moduletag :tmp_dir
 
+  defp ends_with_short_hash?(string) do
+    string
+    |> String.slice(-9..-1)
+    |> String.starts_with?("/")
+  end
+
+  defp ends_with_short_hash_and_extra_path?(string, extra_path) do
+    extra_path = "/" <> extra_path
+    extra_path_length = String.length(extra_path)
+
+    case String.split_at(string, -extra_path_length) do
+      {tmp_dir_base, extra_path_new} when extra_path_new == extra_path ->
+        ends_with_short_hash?(tmp_dir_base)
+
+      _ ->
+        false
+    end
+  end
+
+  defp starts_with_path?(tmp_dir, path) do
+    String.starts_with?(tmp_dir, Path.expand(path))
+  end
+
   test "default path", context do
-    assert context.tmp_dir == Path.expand("tmp/ExUnit.CaseTest.TmpDir/test-default-path")
+    assert starts_with_path?(context.tmp_dir, "tmp/ExUnit.CaseTest.TmpDir/test-default-path/") ==
+             true
+
+    assert ends_with_short_hash?(context.tmp_dir) == true
     assert File.ls!(context.tmp_dir) == []
   end
 
   test "escapes foo?/0", context do
-    assert context.tmp_dir == Path.expand("tmp/ExUnit.CaseTest.TmpDir/test-escapes-foo--0")
+    assert starts_with_path?(context.tmp_dir, "tmp/ExUnit.CaseTest.TmpDir/test-escapes-foo--0/") ==
+             true
+
+    assert ends_with_short_hash?(context.tmp_dir) == true
   end
 
   @tag tmp_dir: "foo/bar"
   test "custom path", context do
-    assert context.tmp_dir == Path.expand("tmp/ExUnit.CaseTest.TmpDir/test-custom-path/foo/bar")
+    assert starts_with_path?(
+             context.tmp_dir,
+             "tmp/ExUnit.CaseTest.TmpDir/test-custom-path/"
+           ) == true
+
+    assert ends_with_short_hash_and_extra_path?(context.tmp_dir, "foo/bar") == true
+  end
+
+  test "colliding-test-names", context do
+    assert starts_with_path?(
+             context.tmp_dir,
+             "tmp/ExUnit.CaseTest.TmpDir/test-colliding-test-names/"
+           ) == true
+
+    assert String.ends_with?(context.tmp_dir, "/e09bee07") == true
+  end
+
+  test "colliding+test+names", context do
+    assert starts_with_path?(
+             context.tmp_dir,
+             "tmp/ExUnit.CaseTest.TmpDir/test-colliding-test-names/"
+           ) == true
```

Comment:
You don't need `== true`. I am almost sure it will actually lead to worse error messages. :D

Reply:
For some reason I had the impression it was the other way around.
Fixed.